github/github.go

```diff
@@ -2,121 +2,193 @@ package github
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
-	"github.com/dgrijalva/jwt-go"
+	"net/url"
+	"strings"
+
 	"github.com/kintohub/utils-go/klog"
 	"github.com/valyala/fasthttp"
-	"strings"
-	"time"
 )
 
 type GithubInterface interface {
-	CallGithub(endpoint, verb, params, authToken string) ([]byte, error)
-	GenerateInstallationToken(installationId string) (string, error)
-	GenBearer(jwtToken string) string
+	GetListRepos(page int32, installationId, userAccessToken string) (*GithubRepositories, error)
+	GetUserInformation(userAccessToken string) (*GithubUserInfo, error)
+	CreateGithubAppToken(installationId string) (string, error)
+	GetUserAccessToken(code string) (string, error)
 }
 
 type github struct {
-	baseUrl       string
-	acceptHeader  string
-	appID         string
-	appPrivateKey []byte
+	appClientID     string
+	appClientSecret string
 }
 
-func New(baseUrl, acceptHeader, appID string, appPrivateKey []byte) GithubInterface {
+var (
+	BASE_API_URL             = "https://api.github.com"
+	BASE_URL                 = "https://github.com"
+	TEMP_ACCEPT_HEADER_VALUE = "application/vnd.github.machine-man-preview+json"
+)
+
+func New(appClientID, appClientSecret string) GithubInterface {
 	g := &github{
-		baseUrl:       baseUrl,
-		acceptHeader:  acceptHeader,
-		appID:         appID,
-		appPrivateKey: appPrivateKey,
+		appClientID:     appClientID,
+		appClientSecret: appClientSecret,
 	}
 	return g
 }
 
-// A function that calls the github api, uses github base url
-// accepts an endpoint that starts with "/" (ex: /app/installation)
-// accept query that will be set after "?" (ex: page=2&per_page=50)
-// sets the default "Accept" header to the one used by github apps
-// authToken is the full auth token not just the value (ex: authToken="Bearer {token}")
-func (g *github) CallGithub(endpoint, verb, query, authToken string) ([]byte, error) {
-	fullUrl := fmt.Sprintf("%s/%s?%s",
-		g.baseUrl, strings.TrimPrefix(endpoint, "/"), strings.TrimPrefix(query, "?"))
-	klog.Debugf("Full Github URL: %v, Auth Token: %v", fullUrl, authToken)
-	req := fasthttp.AcquireRequest()
-	resp := fasthttp.AcquireResponse()
-	defer fasthttp.ReleaseRequest(req)
-	defer fasthttp.ReleaseResponse(resp)
+type githubUserAccessRequest struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+	Code         string `json:"code"`
+}
```

Comment:
comment - why not move that into the `GetUserAccessToken` function since it's only used there?